src/core/processing/qgsprocessingutils.cpp

```diff
@@ -516,7 +516,7 @@ QgsMapLayer *QgsProcessingUtils::mapLayerFromString( const QString &string, QgsP
     return nullptr;
 
   // prefer project layers
-  if ( context.project() && typeHint == LayerHint::Annotation && string.compare( QLatin1String( "main" ), Qt::CaseInsensitive ) == 0 )
+  if ( context.project() && ( typeHint == LayerHint::Annotation || typeHint == LayerHint::UnknownType ) && string.compare( QLatin1String( "main" ), Qt::CaseInsensitive ) == 0 )
```

Comment:
Wouldn't this break things if the project eg contains a vector layer called "main"?

Reply:
Yes, it will break in that case and not only with vector layer but also with other layer types. We probably need to find another way to check validity of main annotation layer here.

Reply:
Ok -- how about splitting this logic then, and leaving the condition where the typeHint is set where it is (ie it gets first preference), and then adding the new logic where typeHint isn't set to just before the "if ( !allowLoadingNewLayers )" check?

That way the named project layer will still get preference when typeHint isn't set.